### PR TITLE
Support setting font weight

### PR DESCRIPTION
**Pull request by nikonso**
Implementation of the feature requested in #1117 based on the comments in the issue.

It seems to be working, but I'm not sure if i missed something.

Should I add typings test for all the available options?

### CODE REVIEW

**Comment by Tyriar** [comment]
Bold doesn't seem to be working on the demo initially:

image

Expected (`v3`):

image

File: src/Terminal.ts [hunk 1]
```diff
@@ -414,14 +415,19 @@ export class Terminal extends EventEmitter implements ITerminal, IInputHandlingT
...
         break;
-      case 'enableBold':
       case 'letterSpacing':
       case 'lineHeight':
         // When the font changes the size of the cells may change which requires a renderer clear
         this.renderer.clear();
         this.renderer.onResize(this.cols, this.rows, false);
         this.refresh(0, this.rows - 1);
         // this.charMeasure.measure(this.options);
+        case 'fontWeight':
```
Comment (Tyriar) [inline src/Terminal.ts]: Fix indentation

File: typings/xterm.d.ts [hunk 2]
```diff
@@ -443,6 +443,12 @@ declare module 'xterm' {
      */
     setOption(key: 'fontFamily' | 'termName' | 'bellSound', value: string): void;
     /**
+    * Sets an option on the terminal.
+    * @param key The option key.
+    * @param value The option value.
+    */
+    setOption(key: 'fontWeight' | 'fontWeightBold', value: null | 'normal' | 'bold' | 'bolder' | 'lighter' | '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900'): void;
```
Comment (Tyriar) [inline typings/xterm.d.ts]: This is great :smile:

I don't think null is a valid value? Maybe we need to validate it in `setOption`?

File: src/Terminal.ts [hunk 3]
```diff
@@ -414,14 +415,19 @@ export class Terminal extends EventEmitter implements ITerminal, IInputHandlingT
...
+        case 'fontWeightBold':
+          // When the font weight changes the size of the cells may change which requires a renderer clear
```
Comment (Tyriar) [inline src/Terminal.ts]: We could merge this and the lineHeight/letterSpacing case and do something like this:

```js
const didCharSizeChange = (key === 'fontWeight' || key === 'fontWeightBold');
this.renderer.onResize(this.cols, this.rows, didCharSizeChange);
```

**Comment by nikonso** [comment]
Locally, and on windows, it seems to be working for me.

image

Do you have any clues as to how i could reproduce the issue? 😕

File: typings/xterm.d.ts (hunk 2 above)
Comment (mofux) [reply typings/xterm.d.ts]: @Tyriar From reading the code, passing `null` will use the default font weight, which seems to be a valid scenario?

File: src/renderer/CharAtlas.ts [hunk 4]
```diff
@@ -137,7 +143,7 @@ class CharAtlasGenerator {
 
     this._ctx.save();
     this._ctx.fillStyle = foreground;
-    this._ctx.font = `${fontSize * window.devicePixelRatio}px ${fontFamily}`;
+    this._ctx.font = `${fontWeight} ${fontSize * window.devicePixelRatio}px ${fontFamily}`;
```
Comment (mofux) [inline src/renderer/CharAtlas.ts]: Would be nice if we would reuse the _getFont method introduced in BaseRenderLayer.

File: src/renderer/CharAtlas.ts [hunk 5]
```diff
@@ -151,7 +157,7 @@ class CharAtlasGenerator {
     }
     // Default color bold
     this._ctx.save();
-    this._ctx.font = `bold ${this._ctx.font}`;
+    this._ctx.font = `${fontWeightBold} ${fontSize * window.devicePixelRatio}px ${fontFamily}`;
```
Comment (mofux) [inline src/renderer/CharAtlas.ts]: Would be nice if we would reuse the _getFont method introduced in BaseRenderLayer.

File: src/renderer/CharAtlas.ts [hunk 6]
```diff
@@ -163,11 +169,11 @@ class CharAtlasGenerator {
     this._ctx.restore();
 
     // Colors 0-15
-    this._ctx.font = `${fontSize * window.devicePixelRatio}px ${fontFamily}`;
+    this._ctx.font = `${fontWeight} ${fontSize * window.devicePixelRatio}px ${fontFamily}`;
```
Comment (mofux) [inline src/renderer/CharAtlas.ts]: Would be nice if we would reuse the _getFont method introduced in BaseRenderLayer.

File: src/renderer/CharAtlas.ts [hunk 7]
```diff
@@ -163,11 +169,11 @@ class CharAtlasGenerator {
...
     for (let colorIndex = 0; colorIndex < 16; colorIndex++) {
       // colors 8-15 are bold
       if (colorIndex === 8) {
-        this._ctx.font = `bold ${this._ctx.font}`;
+        this._ctx.font = `${fontWeightBold} ${fontSize * window.devicePixelRatio}px ${fontFamily}`;
```
Comment (mofux) [inline src/renderer/CharAtlas.ts]: Would be nice if we would reuse the _getFont method introduced in BaseRenderLayer.

File: src/renderer/TextRenderLayer.ts [hunk 8]
```diff
@@ -33,7 +33,7 @@ export class TextRenderLayer extends BaseRenderLayer {
     super.resize(terminal, dim, charSizeChanged);
 
     // Clear the character width cache if the font or width has changed
-    const terminalFont = `${terminal.options.fontSize * window.devicePixelRatio}px ${terminal.options.fontFamily}`;
+    const terminalFont = `${terminal.options.fontWeight} ${terminal.options.fontSize * window.devicePixelRatio}px ${terminal.options.fontFamily}`;
```
Comment (mofux) [inline src/renderer/TextRenderLayer.ts]: Would be nice if we would reuse the _getFont method introduced in BaseRenderLayer.

**Comment by mofux** [comment]
> Bold doesn't seem to be working on the demo initially

Are we seeing a bug here? If I read this comment correctly, bold is only working for normal colored chars, but not for colored chars, which also explains why @nikonso does see it working (white bold text), but @Tyriar is not (colored bold prompt)...

https://github.com/xtermjs/xterm.js/issues/1117#issuecomment-343180815

**Comment by Tyriar** [comment]
@mofux yeah looks like non-white colors in the char atlas are not being drawn as bold:

screen shot 2017-11-09 at 10 17 15 am

**Comment by nikonso** [comment]
Ok, I'l be away from the computer during the weekend, so I'll try to fix
that when I get back on Sunday :wink:

**Comment by Tyriar** [comment]
@nikonso no problems, thanks for all the help recently 😃

**Comment by Tyriar** [comment]
Sorry about the delay, looks good to push in now :+1:

**Pull request merged**